### ISSUE TITLE
Travis auto-checking version increment on PR

### ISSUE THREAD

**Issue by Wolvan** [new-xkit/XKit#768]
It happened to us more than once that we forgot to increment the version on PRs or there have been version changes we didn't realize happened. Maybe we should consider Travis to compare versions in PRs that have to do with extensions

**Comment by dlmarquis** [comment]
I'm for this because extensions need version bumps when edited. We do it manually now, but it's something that doesn't really need human intervention, so why not automate it?

**Comment by Wolvan** [comment]
It's not about auto-changing the version. Autochanging wouldn't really work since the script would have to determine if it's a major, minor or patch bump. This suggestion just says should we automatically let travis make sure that the version on a PR IS bumped

**Comment by dlmarquis** [comment]
Yeah, I meant that we currently have to wait for one of us to check if it's been bumped, which we don't really need a human for. I realize now that I left out the part of that post that made that clear.

**Comment by ThePsionic** [comment]
:+1: do more work to get us to do less work

**Comment by dlmarquis** [comment]
^^Homu also identifies what PR broke the merge, which I find useful.

Yes, this will add One More Thing to check if/when we implement either of those changes. However, I don't see why this is a prohibitive concern: either we remember immediately, or we remember when someone makes a PR and Travis breaks. Worst case, we have to fix version numbers before merging our beta stuff to master, since we have to manually check that the version was incremented properly before the release anyway, we shouldn't accidentally merge anything with too high a version number. Neither senario will cause horrible reprocussions, in any case.

Regarding our current review issues, this isn't supposed to be a quick fix for our PR-handling problems. Even if we got our shit together, a pre-check for something easily forgotten would be useful. While we definitely have an issue with not checking PRs carefully, we also have a habit of forgetting about older PRs that will not be solved by creating a standard checklist. It's due to all of us having day jobs and real life stuff, and I don't forsee us being able to eliminate it entirely. Therefore, we should be looking for ways to get the easy stuff checked and fixed before someone reviews it. A version increase check in Travis would help here.

I don't understand your concern with adding a different type of feature to Travis. No, version checking doesn't fall under any of the current types of check, but why does that really matter? Travis's usefulness lies in its checking for menial things like linting that we shouldn't have to waste time doing manually, or risk the PR getting cordoned off into purgatory as I described above. Not using it to its full potential because of concern over what it's "supposed to" contain just seems silly to me.

As for version-bumping in PRs, we're in a unique position where we can't incubate things in master between builds. There's very little delay between when something gets to master and when it reaches the user, even without a version increase. Updating through the My XKit panel will grab whatever's on master, regardless of the version, and if we leave things long enough for people to do that, we'll have a support nightmare where the version means nothing. So if we don't have version incrementing in the PR, we'd have to do it immediately after the PR is merged, solving none of our merge conflict problems.

Forcing people to manually check that the version was incremented isn't going to make us any better at checking the body of the PR any more than making it automatic is going to suddenly result in people not caring to test anything locally before merging. It will just be one more thing that people may or may not remember to look at. What will do what you're looking for is a review guide, which can easily coexist with a Travis check, as I mentioned earlier. Further, we'll still need to check that the version was updated correctly, which is easier to remember when the relevant line pops up in the changes tab.

Generally, I don't see the downsides of this being great enough to forgo doing it, and I don't think it will have an effect either way on the thouroughness of review.

**Comment by Wolvan** [comment]
Small correction here, @dlmarquis: Updating through the panel does NOT grab whatever is on master, it does compare versions. The only extensions that get fetched on forced update, regardless of version, are the 4 internals

**Comment by dlmarquis** [comment]
@Wolvan I use it to grab non-incremented versions all the time. Unless it's meant to work differently on local host than on Github?

I mean the update button when looking at a specific extension. I know the main force update button checks versions.

**Comment by Wolvan** [comment]
Oh I see! Sorry for misunderstanding

**Comment by 0xazure** [comment]
> protected branches weren't a thing when homu was released

You're correct, @nightpool, but that's no reason not to reconsider our tools to achieve the same result.

As for @dlmarquis' comments:

> Homu also identifies what PR broke the merge, which I find useful.

Would you indulge me, I'd be interested to hear your approach to reviewing PRs because it seems we're leveraging tools in different ways.  I will grant you, Homu is useful to know if a PR I have submitted is out of date with master, but that's all I use it for.  If I want to know _why_ the PR is out of date, I start reaching for `git diff` et al., primarily because Homu only alerts on the first conflict.  

Here's a scenario: you make PR 6, and Homu comments that PR 8 made 6 unmergeable.  Between the time of that alert and when you have time to update PR 6, PRs 12, 16, and 20 have been merged, all of which also conflict with your changes.  Knowing that PR 8 made 6 unmergeable doesn't help when there are 3 other PRs whose conflicts you're going to have to resolve.  This scenario is not unfounded given how long it sometimes takes us to review PRs (again, see #554).

> merge anything with too high a version number

I'm not sure I understand this, could you clarify?

> we also have a habit of forgetting about older PRs that will not be solved by creating a standard checklist

Why would a standard checklist not help older PRs?

> we have to fix version numbers before merging our beta stuff to master

This is actually an interesting point to consider.  How would we deal with version conflicts between stable and beta?  Must beta be continuously "ahead" of stable (i.e. if stable updates from `4.0.2` to `4.1.0`, will beta have to update to `4.2.0` (assuming a `MINOR` increment) before it can be merged anywhere?  Are we limiting this check only to PRs targeting `master`, or will it apply to _all_ of our branches?  How will this relate to the beta release channel?

> I don't understand your concern with adding a different type of feature to Travis.

My fault, I probably should have expanded on this.  My concern is in separating the signal from the noise.  If we add this kind of checking to TravisCI, we'll see the build fail, sure.  However, tracking down the _reason_ for the failure becomes more involved the more tasks we add to a single TravisCI test.  I know from experience that (visually) `grep`ping CI logs is annoying at the best of times, and confusing at the worst.  It also unreasonable to expect contributors from the community to become intimately familiar with our build/test process and TravisCI just to contribute to the project.

I would be less opposed to implementing this kind of checking in a similar manner to [GitCop](https://gitcop.com/home) and [Coveralls](https://coveralls.io/) checks (among others), whose services leverage the [GitHub webhooks infrastructure](https://developer.github.com/webhooks/) to provide comments on issues when checks fail.  This would go a long way to making it easy for contributors to see that they have forgotten to make whatever changes are required  without having to wade through CI logs.  It would also be extensible, should there be other checks we want to add in the future.

> we can't incubate things in master between builds

This is simply due to the fact that we implemented automatic distribution re-building once the TravisCI checks on `master` pass, and has nothing to do with any particular requirement of our release schedule (or lack thereof).

> So if we don't have version incrementing in the PR, we'd have to do it immediately after the PR is merged

I suppose I wasn't clear here either, my apologies.  I said nothing about incrementing versions _after_ PRs get merged.  Instead, for example, a Collaborator would take the contents of a submitted PR, add an additional commit to the `HEAD` of that branch which properly increments the version(s), and then merge the branch into `master` on behalf of the submitter (contribution details would be preserved, and GitHub respects the difference between "committer" and "author").  Yes, this is more involved than simply clicking on a button on the web UI, but I see that as a positive.  By requiring the reviewer to check out the relevant branch to approve the changes, it significantly lowers the barrier to actively manually testing the changeset (which, given some turn-arounds on previous PRs, I know for a fact does not happen as often as it should).

> we'll still need to check that the version was updated correctly

If we still have to check that the version was updated correctly, we have to look for a version increment in the first place.

> a review guide, which can easily coexist with a [...] check

You're not wrong in this, and as I'd mentioned previously, a formalized process would go a long way.

**Comment by bvtsang** [comment]
> Here's a scenario: you make PR 6, and Homu comments that PR 8 made 6 unmergeable. Between the time of that alert and when you have time to update PR 6, PRs 12, 16, and 20 have been merged, all of which also conflict with your changes. Knowing that PR 8 made 6 unmergeable doesn't help when there are 3 other PRs whose conflicts you're going to have to resolve.

Correct me if I'm wrong, but in this scenario knowing that PR 8 made 6 unmergeable still helps in resolving conflicts with PRs 12, 16, and 20.

Let's say Alice's git history looks like the following (most to least recent commits):

```
* guid6 - PR 6 (feature branch)
* guid3 - C (local master)
* guid2 - B
* guid1 - A
```

And let's say that New XKit's `master` branch looks like:

```
* guid20 - PR 20 (upstream master)
* guid16 - PR 16
* guid12 - PR 12
* guid8 - PR 8
* guid3 - C
* guid2 - B
* guid1 - A
```

Homu has just told Alice that PR 8 conflicts with PR 6. Naturally, in order to resolve this, Alice fetches the contents of upstream `master` so she can begin to see what she needs to do. And in the process she sees this:

```
* guid20 - PR 20 (upstream master)
* guid16 - PR 16
* guid12 - PR 12
* guid8 - PR 8
| * guid6 - PR 6 (feature branch)
|/
* guid3 - C (local master)
* guid2 - B
* guid1 - A
```

It's at that moment that Alice realizes that PRs 8, 12, 16, and 20 may all conflict with PR 6. She knows this because she wants PR 6 to be the most recent commit in the history, and in doing that she needs to either rebase on top of or merge with `master`, which requires reconciling PR 6 with everything between local `master` and upstream `master`.

**Comment by 0xazure** [comment]
@bvtsang You're entirely correct.  PRs 12, 16, and 20 _may_ conflict with PR 6.  But, as you said yourself:

> Alice fetches the contents of upstream master so she can begin to see what she needs to do.

All Homu alerted on was PR 8.  And I'm not saying that's not useful information to have (particularly since it gets propagated through emails and GitHub's notification system).  What I _am_ saying is that Homu does not then subsequently alert that PRs 12, 16, and 20 _also_ make PR 8 unmergable (which I don't think would be possible to do, actually, but that's besides the point).  It isn't until Alice starts reaching for `git` and other tooling outside of Homu's comments that she would notice that there are more conflicts to resolve than those introduced by PR 8.

But these are just my opinions, based on my existing workflow.  I have no idea how anyone else is using the tooling we have set up, which is why concrete examples of tool usefulness is imperative for making a decision here.

**Comment by bvtsang** [comment]
@0xazure Ah, so the difference in meaning is knowing that PR 8 plus potentially other PRs need to be resolved versus knowing _exactly_ that PRs 8, 12, 16, and 20 need to be resolved. Thanks for the clarification!

> I have no idea how anyone else is using the tooling we have set up, which is why concrete examples of tool use-cases are imperative for making a decision here.

To provide a data point, I think it's a good time to describe how I work with Github. The usual disclaimers apply here: this is my process, I'm not sure if it's something everyone else does, etc.

1. Before making a pull request, update and rebase with upstream repository
2. Make a pull request
3. As the pull request receives feedback, make more local commits
4. Before pushing again, always update with upstream repository
5. Loop back to step 3 as needed
6. When the pull request is merged, delete my feature branch from Github and local repository

I've always made an effort to double-check that I'm up to date before pushing my changes. I can recall one recent incident where I've had to fix the version number of Editable Reblogs to account for upstream changes (https://github.com/new-xkit/XKit/pull/763). This version fix was not immediately clear to other people, [especially reviewers](https://gitter.im/new-xkit/XKit?at=5647ac488b242470793dbb96).

----

There's a lot of information to absorb in this thread, so I'd like to summarize what I've read so far.

For:

- Humans forget to bump versions; machines don't
- It's simple and easy
- It would be a waste not to use Travis to its fullest potential
- Especially imperative because merges to `master` make updates immediately available to public

Against:

- More maintenance for Travis scripts
- This is just a bandaid on a symptom of a larger problem: lack of formal review process
- Conflates what Travis is checking (not a lint, code style check, or unit test)
- Adding more tasks to Travis makes it harder to visually grep for the cause of failures

Open questions:

- How would we deal with version conflicts between stable and beta?
- Must beta be continuously "ahead" of stable (i.e. if stable updates from 4.0.2 to 4.1.0, will beta have to update to 4.2.0 (assuming a MINOR increment) before it can be merged anywhere?
- Are we limiting this check only to PRs targeting `master`, or will it apply to all of our branches? How will this relate to the beta release channel?
